From 0e2abc407ad0ff3bf668e3592f7e4397b88c3998 Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sat, 10 Apr 2021 12:02:19 -0500
Subject: [PATCH 072/196] Add Taproot key spend path (#130)

This allows Taproot outputs to be spent with a signature for the commitment using BIP341 signatures.

They are still non-standard, but will become standard in a future commit
---
 src/script/interpreter.cpp                    |  29 +-
 src/script/script_error.cpp                   |   4 +
 src/script/script_error.h                     |   4 +
 src/script/taproot.h                          |   2 +
 src/test/data/script_tests.json               |  10 +-
 src/test/script_tests.cpp                     |   3 +
 src/test/taproot_tests.cpp                    | 218 ++++++++++++++-
 .../logos_feature_taproot_key_spend.py        | 251 ++++++++++++++++++
 8 files changed, 516 insertions(+), 5 deletions(-)
 create mode 100644 test/functional/logos_feature_taproot_key_spend.py

diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index 8d54791fa..f1618852e 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -2051,7 +2051,34 @@ static bool VerifyTaprootSpend(std::vector<valtype> stack,
     if (!IsPayToTaproot(script_pubkey)) {
         return set_error(serror, ScriptError::SCRIPTTYPE_MALFORMED_SCRIPT);
     }
-    // Temporarily return UNKNOWN
+    valtype vch_pubkey =
+        valtype(script_pubkey.begin() + TAPROOT_INTRO_SIZE,
+                script_pubkey.begin() + TAPROOT_SIZE_WITHOUT_STATE);
+
+    if (stack.size() == 0) {
+        return set_error(serror, ScriptError::INVALID_STACK_OPERATION);
+    }
+    if (stack.size() >= 2 && !stack.back().empty() &&
+        stack.back()[0] == TAPROOT_ANNEX_TAG) {
+        return set_error(serror, ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED);
+    }
+    if (stack.size() == 1) {
+        // Spend using single signature instead of executing script
+        const valtype &vch_sig = stacktop(-1);
+        const uint32_t sig_flags = flags | SCRIPT_TAPROOT_KEY_SPEND_PATH;
+        if (!CheckTransactionSignatureEncoding(vch_sig, sig_flags, serror) ||
+            !CheckPubKeyEncoding(vch_pubkey, sig_flags, serror)) {
+            // serror is set
+            return false;
+        }
+        if (vch_sig.empty() || !checker.CheckSig(vch_sig, vch_pubkey,
+                                                 std::nullopt, {}, sig_flags)) {
+            return set_error(serror,
+                             ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED);
+        }
+        return set_success(serror);
+    }
+    // Temporarily return UNKNOWN for script spend path
     return set_error(serror, ScriptError::UNKNOWN);
 }
 
diff --git a/src/script/script_error.cpp b/src/script/script_error.cpp
index 73c7a9782..9df5d57fe 100644
--- a/src/script/script_error.cpp
+++ b/src/script/script_error.cpp
@@ -118,6 +118,10 @@ std::string ScriptErrorString(const ScriptError serror) {
             return "Taproot key spend signatures must use SIGHASH_BIP341";
         case ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG:
             return "Taproot key spend signature must be Schnorr";
+        case ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED:
+            return "Invalid Taproot key spend signature";
+        case ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED:
+            return "BIP341-style taproot annex (0x50) not supported";
         case ScriptError::UNKNOWN:
         case ScriptError::ERROR_COUNT:
         default:
diff --git a/src/script/script_error.h b/src/script/script_error.h
index d157f7a32..c32391b50 100644
--- a/src/script/script_error.h
+++ b/src/script/script_error.h
@@ -97,6 +97,10 @@ enum class ScriptError {
     TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH,
     /* Key spend path for Taproot must Schnorr signatures */
     TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG,
+    /* Key spend path for Taproot failed */
+    TAPROOT_VERIFY_SIGNATURE_FAILED,
+    /* Taproot annex not supported */
+    TAPROOT_ANNEX_NOT_SUPPORTED,
 
     ERROR_COUNT,
 };
diff --git a/src/script/taproot.h b/src/script/taproot.h
index c12eaca0c..f4ddfd5d6 100644
--- a/src/script/taproot.h
+++ b/src/script/taproot.h
@@ -28,6 +28,8 @@ static constexpr uint32_t TAPROOT_SIZE_WITHOUT_STATE =
 static constexpr uint32_t TAPROOT_SIZE_WITH_STATE =
     TAPROOT_INTRO_SIZE + CPubKey::COMPRESSED_SIZE + 33;
 
+static constexpr uint32_t TAPROOT_ANNEX_TAG = 0x50;
+
 /**
  * Verifies that the control block proves that script is part of the commitment.
  *
diff --git a/src/test/data/script_tests.json b/src/test/data/script_tests.json
index 9eb91db6c..506b0b73d 100644
--- a/src/test/data/script_tests.json
+++ b/src/test/data/script_tests.json
@@ -1849,7 +1849,7 @@
 ["","SCRIPTTYPE 1 0x22 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123", "",
  "SCRIPTTYPE_MALFORMED_SCRIPT", "Taproot script commitment too long"],
 ["","SCRIPTTYPE 1 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01", "",
- "UNKNOWN_ERROR", "Taproot script valid but currently disabled"],
+ "INVALID_STACK_OPERATION", "Taproot script valid but empty scriptSig"],
 ["","SCRIPTTYPE 1 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01 0", "",
  "SCRIPTTYPE_MALFORMED_SCRIPT", "Taproot script state zero"],
 ["","SCRIPTTYPE 1 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01 0x05 0x0102030405", "",
@@ -1857,9 +1857,15 @@
 ["","SCRIPTTYPE 1 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01 0x1f 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcd", "",
  "SCRIPTTYPE_MALFORMED_SCRIPT", "Taproot script state too short"],
 ["","SCRIPTTYPE 1 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01 0x20 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef", "",
- "UNKNOWN_ERROR", "Taproot script valid but currently disabled"],
+ "INVALID_STACK_OPERATION", "Taproot script valid but empty scriptSig"],
 ["","SCRIPTTYPE 1 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01", "",
  "SCRIPTTYPE_MALFORMED_SCRIPT", "Taproot script state too long"],
+["0 0x01 0x50","SCRIPTTYPE 1 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01", "",
+ "TAPROOT_ANNEX_NOT_SUPPORTED", "Annex not supported"],
+["0x10 0x0123456789abcdef0123456789abcdef 0x01 0x50","SCRIPTTYPE 1 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01", "",
+ "TAPROOT_ANNEX_NOT_SUPPORTED", "Annex not supported"],
+["0x10 0x0123456789abcdef0123456789abcdef 0x03 0x500101","SCRIPTTYPE 1 0x21 0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef01", "",
+ "TAPROOT_ANNEX_NOT_SUPPORTED", "Annex not supported"],
 
 ["1","VERIF", "", "DISABLED_OPCODE", "OP_VERIF is reserved"],
 ["1","VERNOTIF", "", "DISABLED_OPCODE", "OP_VERNOTIF is reserved"],
diff --git a/src/test/script_tests.cpp b/src/test/script_tests.cpp
index 7914f393f..48027df80 100644
--- a/src/test/script_tests.cpp
+++ b/src/test/script_tests.cpp
@@ -101,6 +101,9 @@ static ScriptErrorDesc script_errors[] = {
      "TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG"},
     {ScriptError::SCRIPTTYPE_INVALID_TYPE, "SCRIPTTYPE_INVALID_TYPE"},
     {ScriptError::SCRIPTTYPE_MALFORMED_SCRIPT, "SCRIPTTYPE_MALFORMED_SCRIPT"},
+    {ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED,
+     "TAPROOT_VERIFY_SIGNATURE_FAILED"},
+    {ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED, "TAPROOT_ANNEX_NOT_SUPPORTED"},
 };
 
 static std::string FormatScriptError(ScriptError err) {
diff --git a/src/test/taproot_tests.cpp b/src/test/taproot_tests.cpp
index 2ed95c281..af8d00e4d 100644
--- a/src/test/taproot_tests.cpp
+++ b/src/test/taproot_tests.cpp
@@ -295,6 +295,24 @@ struct TestUtxo {
     COutPoint outpoint = COutPoint();
 };
 
+struct TestInput {
+    CScript script_sig;
+    CScript script_pubkey;
+    CScript exec_script;
+    Amount amount;
+    SigHashType sig_hash_type;
+    bool use_schnorr;
+    bool is_key_path_spend = true;
+    uint32_t sequence = 0xffff'ffff;
+    COutPoint outpoint = COutPoint();
+    uint32_t codeseparator_pos = 0xffff'ffff;
+    TestUtxo utxo() const {
+        return {
+            script_sig, script_pubkey, exec_script, amount, sequence, outpoint,
+        };
+    }
+};
+
 void CheckTxScripts(const std::vector<TestUtxo> &inputs,
                     const std::vector<CTxOut> &outputs, const uint32_t flags,
                     const ScriptError expected_error = ScriptError::OK) {
@@ -325,6 +343,40 @@ void CheckTxScripts(const std::vector<TestUtxo> &inputs,
     }
 }
 
+std::vector<valtype> MakeSigs(const CKey &seckey,
+                              const std::vector<TestInput> &inputs,
+                              const std::vector<CTxOut> &outputs,
+                              const uint32_t flags) {
+    CMutableTransaction tx;
+    std::vector<CTxOut> spent_outputs;
+    for (const TestInput &input : inputs) {
+        tx.vin.push_back(
+            CTxIn(input.outpoint, input.script_sig, input.sequence));
+        spent_outputs.push_back({input.amount, input.script_pubkey});
+    }
+    tx.vout = outputs;
+    const PrecomputedTransactionData txdata(tx, std::move(spent_outputs));
+    std::vector<valtype> sigs(inputs.size());
+    for (uint32_t input_idx = 0; input_idx < inputs.size(); ++input_idx) {
+        const TestInput &input = inputs[input_idx];
+        std::optional<ScriptExecutionData> execdata;
+        assert(input.is_key_path_spend);
+        const CScript script_code = CScript();
+        uint256 sighash;
+        BOOST_CHECK(SignatureHash(
+            sighash, execdata, script_code, tx, input_idx, input.sig_hash_type,
+            txdata.m_spent_outputs[input_idx].nValue, &txdata, flags));
+        if (input.use_schnorr) {
+            BOOST_CHECK(seckey.SignSchnorr(sighash, sigs[input_idx]));
+        } else {
+            BOOST_CHECK(seckey.SignECDSA(sighash, sigs[input_idx]));
+        }
+        sigs[input_idx].push_back(input.sig_hash_type.getRawSigHashType() &
+                                  0xff);
+    }
+    return sigs;
+}
+
 BOOST_AUTO_TEST_CASE(verify_invalid_taproot_script) {
     valtype vch_pubkey = ParseHex(
         "020000000000000000000000000000000000000000000000000000000000000001");
@@ -354,9 +406,9 @@ BOOST_AUTO_TEST_CASE(verify_invalid_taproot_script) {
         {{CScript() << OP_SCRIPTTYPE << OP_1 << valtype(32) << valtype(33)},
          ScriptError::SCRIPTTYPE_MALFORMED_SCRIPT},
         {{CScript() << OP_SCRIPTTYPE << OP_1 << valtype(33)},
-         ScriptError::UNKNOWN},
+         ScriptError::INVALID_STACK_OPERATION},
         {{CScript() << OP_SCRIPTTYPE << OP_1 << valtype(33) << valtype(32)},
-         ScriptError::UNKNOWN},
+         ScriptError::INVALID_STACK_OPERATION},
     };
     for (auto &pair : invalid_scripts) {
         std::vector<TestUtxo> inputs = {{{}, pair.first, {}, Amount::zero()}};
@@ -367,4 +419,166 @@ BOOST_AUTO_TEST_CASE(verify_invalid_taproot_script) {
     }
 }
 
+BOOST_AUTO_TEST_CASE(verify_invalid_taproot_spend) {
+    valtype vch_pubkey = ParseHex(
+        "020000000000000000000000000000000000000000000000000000000000000001");
+    std::vector<std::pair<std::pair<stacktype, CScript>, ScriptError>>
+        invalid_scripts = {
+            {{{}, CScript() << OP_SCRIPTTYPE << OP_1 << valtype(33)},
+             ScriptError::INVALID_STACK_OPERATION},
+            {{{{}, {0x50}}, CScript() << OP_SCRIPTTYPE << OP_1 << valtype(33)},
+             ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED},
+            {{{{0x00, 0x00}, {0x50}},
+              CScript() << OP_SCRIPTTYPE << OP_1 << valtype(33)},
+             ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED},
+            {{{{}, {0x50, 0x00, 0x00}},
+              CScript() << OP_SCRIPTTYPE << OP_1 << valtype(33)},
+             ScriptError::TAPROOT_ANNEX_NOT_SUPPORTED},
+            {{{{}}, CScript() << OP_SCRIPTTYPE << OP_1 << valtype(33)},
+             ScriptError::PUBKEYTYPE},
+            {{{{}}, CScript() << OP_SCRIPTTYPE << OP_1 << vch_pubkey},
+             ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED},
+            {{{valtype(65)}, CScript() << OP_SCRIPTTYPE << OP_1 << vch_pubkey},
+             ScriptError::SIG_HASHTYPE},
+            {{{{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x61}},
+              CScript() << OP_SCRIPTTYPE << OP_1 << vch_pubkey},
+             ScriptError::TAPROOT_VERIFY_SIGNATURE_FAILED},
+            {{{{0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+                0x00, 0x00, 0x00, 0x00, 0x65}},
+              CScript() << OP_SCRIPTTYPE << OP_1 << vch_pubkey},
+             ScriptError::SIG_HASHTYPE},
+        };
+    for (auto &pair : invalid_scripts) {
+        CScript script_sig;
+        for (const valtype &item : pair.first.first) {
+            script_sig = script_sig << item;
+        }
+        TestUtxo input = {
+            .script_sig = script_sig,
+            .script_pubkey = pair.first.second,
+            .exec_script = pair.first.second,
+            .amount = Amount::zero(),
+        };
+        const std::vector<CTxOut> outputs = {{Amount::zero(), {}}};
+        for (uint32_t flags : flagset) {
+            CheckTxScripts({input}, outputs, flags, pair.second);
+        }
+    }
+}
+
+BOOST_AUTO_TEST_CASE(verify_key_spend) {
+    CKey seckey;
+    const valtype vch_seckey = ParseHex(
+        "0000000000000000000000000000000000000000000000000000000000000001");
+    seckey.Set(vch_seckey.begin(), vch_seckey.end(), true);
+    CPubKey pubkey = seckey.GetPubKey();
+    const valtype vch_pubkey(pubkey.begin(), pubkey.end());
+    const CScript script = CScript() << OP_SCRIPTTYPE << OP_1 << vch_pubkey;
+    const Amount amount = 1337 * COIN;
+    const SigHashType sighash_all = SigHashType(SIGHASH_ALL).withBIP341();
+    const SigHashType sighash_none = SigHashType(SIGHASH_NONE).withBIP341();
+    const SigHashType sighash_single = SigHashType(SIGHASH_SINGLE).withBIP341();
+    const std::vector<std::vector<TestInput>> success_input_cases = {
+        {{{}, script, {}, amount, sighash_all, true}},
+        {{{}, script, {}, amount, sighash_all.withAnyoneCanPay(), true}},
+        {{{}, script, {}, amount, sighash_none, true}},
+        {{{}, script, {}, amount, sighash_none.withAnyoneCanPay(), true}},
+        {{{}, script, {}, amount, sighash_single, true}},
+        {{{}, script, {}, amount, sighash_single.withAnyoneCanPay(), true}},
+        {{{}, script, {}, amount, sighash_single, true},
+         {{}, script, {}, amount, sighash_all, true},
+         {{}, script, {}, amount, sighash_all.withAnyoneCanPay(), true},
+         {{}, script, {}, amount, sighash_none, true},
+         {{}, script, {}, amount, sighash_none.withAnyoneCanPay(), true}},
+    };
+    const std::vector<std::pair<std::vector<TestInput>, ScriptError>>
+        error_input_cases = {
+            {{{{}, script, {}, amount, sighash_all, false}},
+             ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG},
+            {{{{}, script, {}, amount, sighash_all.withAnyoneCanPay(), false}},
+             ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG},
+            {{{{}, script, {}, amount, sighash_none, false}},
+             ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG},
+            {{{{}, script, {}, amount, sighash_none.withAnyoneCanPay(), false}},
+             ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG},
+            {{{{}, script, {}, amount, sighash_single, false}},
+             ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG},
+            {{{{},
+               script,
+               {},
+               amount,
+               sighash_single.withAnyoneCanPay(),
+               false}},
+             ScriptError::TAPROOT_KEY_SPEND_MUST_USE_SCHNORR_SIG},
+            {{{{},
+               script,
+               {},
+               amount,
+               sighash_all.withAlgorithm(SIGHASH_LEGACY),
+               true}},
+             ScriptError::TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH},
+            {{{{},
+               script,
+               {},
+               amount,
+               sighash_all.withAlgorithm(SIGHASH_FORKID),
+               true}},
+             ScriptError::TAPROOT_KEY_SPEND_MUST_USE_BIP341_SIGHASH},
+        };
+    const std::vector<std::vector<CTxOut>> output_cases = {
+        {{Amount::zero(), {}}},
+        {{10 * COIN, CScript() << OP_1}},
+        {{10 * COIN, CScript() << OP_1},
+         {20 * COIN, CScript() << OP_3},
+         {30 * COIN, CScript() << OP_4}}};
+    std::vector<std::pair<std::vector<TestInput>, ScriptError>> input_cases;
+    for (const auto &inputs : success_input_cases) {
+        input_cases.push_back({inputs, ScriptError::OK});
+    }
+    input_cases.insert(input_cases.end(), error_input_cases.begin(),
+                       error_input_cases.end());
+    for (const auto &pair : input_cases) {
+        const std::vector<TestInput> &inputs = pair.first;
+        const ScriptError serror = pair.second;
+        std::vector<TestUtxo> signed_inputs(inputs.size());
+        for (const std::vector<CTxOut> &outputs : output_cases) {
+            for (uint32_t flags : flagset) {
+                std::vector<valtype> sigs =
+                    MakeSigs(seckey, inputs, outputs, flags);
+                for (uint32_t i = 0; i < inputs.size(); ++i) {
+                    signed_inputs[i] = inputs[i].utxo();
+                    signed_inputs[i].script_sig = CScript() << sigs[i];
+                }
+                CheckTxScripts(signed_inputs, outputs, flags, serror);
+            }
+        }
+    }
+}
+
+BOOST_AUTO_TEST_CASE(verify_taproot_simple_contract_disabled) {
+    const std::vector<CTxOut> outputs = {{Amount::zero(), {}}};
+    for (uint32_t flags : flagset) {
+        CheckTxScripts({{CScript() << valtype(0) << valtype(1),
+                         CScript() << OP_SCRIPTTYPE << OP_1 << valtype(33),
+                         CScript(), Amount::zero()}},
+                       outputs, flags, ScriptError::UNKNOWN);
+        CheckTxScripts(
+            {{CScript() << valtype(0) << valtype(1),
+              CScript() << OP_SCRIPTTYPE << OP_1 << valtype(33) << valtype(32),
+              CScript(), Amount::zero()}},
+            outputs, flags, ScriptError::UNKNOWN);
+    }
+}
+
 BOOST_AUTO_TEST_SUITE_END()
diff --git a/test/functional/logos_feature_taproot_key_spend.py b/test/functional/logos_feature_taproot_key_spend.py
new file mode 100644
index 000000000..cbbc9a8d5
--- /dev/null
+++ b/test/functional/logos_feature_taproot_key_spend.py
@@ -0,0 +1,251 @@
+#!/usr/bin/env python3
+# Copyright (c) 2021 The Logos Foundation
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""
+Test that we can broadcast transactions spending a Taproot output using the key
+spend path.
+
+Outputs have the form:
+
+OP_SCRIPTTYPE OP_1 <33-byte commitment pubkey>
+or
+OP_SCRIPTTYPE OP_1 <33-byte commitment pubkey> <32-byte state>
+
+and inputs have the form:
+
+<65-byte Schnorr signature using BIP341>
+"""
+
+from test_framework.key import ECKey
+from test_framework.blocktools import create_block, create_coinbase, make_conform_to_ctor
+from test_framework.messages import (
+    COutPoint,
+    CTransaction,
+    CTxIn,
+    CTxOut,
+)
+from test_framework.mininode import (
+    P2PDataStore,
+)
+from test_framework.script import (
+    CScript,
+    OP_1,
+    OP_EQUAL,
+    OP_HASH160,
+    OP_SCRIPTTYPE,
+    SIGHASH_ALL,
+    SIGHASH_ANYONECANPAY,
+    SIGHASH_BIP341,
+    SIGHASH_FORKID,
+    SIGHASH_NONE,
+    SIGHASH_SINGLE,
+    SignatureHashBIP341,
+)
+from test_framework.test_framework import BitcoinTestFramework
+
+
+TESTCASES = [
+    dict(outputs=1, inputs=1, sig_hash_types=[SIGHASH_ALL | SIGHASH_BIP341]),
+    dict(outputs=10, inputs=1, sig_hash_types=[SIGHASH_ALL | SIGHASH_BIP341]),
+    dict(outputs=2, inputs=1, sig_hash_types=[SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY]),
+    dict(outputs=3, inputs=1, sig_hash_types=[SIGHASH_NONE | SIGHASH_BIP341]),
+    dict(outputs=4, inputs=1, sig_hash_types=[SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY]),
+    dict(outputs=5, inputs=1, sig_hash_types=[SIGHASH_SINGLE | SIGHASH_BIP341]),
+    dict(outputs=6, inputs=1, sig_hash_types=[SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY]),
+    dict(outputs=10, inputs=6, sig_hash_types=[SIGHASH_ALL | SIGHASH_BIP341,
+                                               SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY,
+                                               SIGHASH_NONE | SIGHASH_BIP341,
+                                               SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY,
+                                               SIGHASH_SINGLE | SIGHASH_BIP341,
+                                               SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY]),
+    dict(outputs=1, inputs=1, sig_hash_types=[SIGHASH_ALL | SIGHASH_BIP341],
+         ecdsa=True, error='Taproot key spend signature must be Schnorr'),
+    dict(outputs=10, inputs=1, sig_hash_types=[SIGHASH_ALL | SIGHASH_BIP341],
+         ecdsa=True, error='Taproot key spend signature must be Schnorr'),
+    dict(outputs=2, inputs=1, sig_hash_types=[SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY],
+         ecdsa=True, error='Taproot key spend signature must be Schnorr'),
+    dict(outputs=3, inputs=1, sig_hash_types=[SIGHASH_NONE | SIGHASH_BIP341],
+         ecdsa=True, error='Taproot key spend signature must be Schnorr'),
+    dict(outputs=4, inputs=1, sig_hash_types=[SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY],
+         ecdsa=True, error='Taproot key spend signature must be Schnorr'),
+    dict(outputs=5, inputs=1, sig_hash_types=[SIGHASH_SINGLE | SIGHASH_BIP341],
+         ecdsa=True, error='Taproot key spend signature must be Schnorr'),
+    dict(outputs=6, inputs=1, sig_hash_types=[SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY],
+         ecdsa=True, error='Taproot key spend signature must be Schnorr'),
+    dict(outputs=10, inputs=6, sig_hash_types=[SIGHASH_ALL | SIGHASH_BIP341,
+                                               SIGHASH_ALL | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY,
+                                               SIGHASH_NONE | SIGHASH_BIP341,
+                                               SIGHASH_NONE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY,
+                                               SIGHASH_SINGLE | SIGHASH_BIP341,
+                                               SIGHASH_SINGLE | SIGHASH_BIP341 | SIGHASH_ANYONECANPAY],
+         ecdsa=True, error='Taproot key spend signature must be Schnorr'),
+    dict(outputs=4, inputs=1, sig_hash_types=[SIGHASH_BIP341], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x20, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x21, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x2f, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x60, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=1, inputs=3, sig_hash_types=[0x63, 0x61, 0x61]),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x61, 0x63]),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x62, 0x63]),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x63, 0x63]),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x64, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x6f, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0x71, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0xe0, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0xe1, 0x63]),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0xe2, 0x63]),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0xe3, 0x63]),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0xe4, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0xef, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=3, sig_hash_types=[0x61, 0xf1, 0x63], error='Signature hash type missing or not understood'),
+    dict(outputs=5, inputs=1, sig_hash_types=[SIGHASH_FORKID], error='Signature hash type missing or not understood'),
+    dict(outputs=4, inputs=1, sig_hash_types=[SIGHASH_ALL],
+         error='Taproot key spend signatures must use SIGHASH_BIP341'),
+    dict(outputs=3, inputs=1, sig_hash_types=[SIGHASH_NONE],
+         error='Taproot key spend signatures must use SIGHASH_BIP341'),
+    dict(outputs=2, inputs=1, sig_hash_types=[SIGHASH_SINGLE],
+         error='Taproot key spend signatures must use SIGHASH_BIP341'),
+    dict(outputs=6, inputs=1, sig_hash_types=[SIGHASH_ALL | SIGHASH_FORKID],
+         error='Taproot key spend signatures must use SIGHASH_BIP341'),
+    dict(outputs=6, inputs=1, sig_hash_types=[SIGHASH_NONE | SIGHASH_FORKID],
+         error='Taproot key spend signatures must use SIGHASH_BIP341'),
+    dict(outputs=6, inputs=1, sig_hash_types=[SIGHASH_SINGLE | SIGHASH_FORKID],
+         error='Taproot key spend signatures must use SIGHASH_BIP341'),
+    dict(outputs=3, inputs=1, sig_hash_types=[0x21], error='Invalid Taproot key spend signature', suffix=0x61),
+    dict(outputs=3, inputs=1, sig_hash_types=[0x81], error='Invalid Taproot key spend signature', suffix=0xe1),
+    dict(outputs=3, inputs=1, sigs=[b''], error='Invalid Taproot key spend signature'),
+    'ENABLE_REPLAY_PROTECTION',
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdead60], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdead61]),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdead62]),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdead63]),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdead64], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdead6f], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdead71], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdeade0], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdeade1]),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdeade2]),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdeade3]),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdeade4], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdeadef], error='Signature hash type missing or not understood'),
+    dict(outputs=3, inputs=1, sig_hash_types=[0xffdeadf1], error='Signature hash type missing or not understood'),
+]
+
+ACTIVATION_TIME = 2000000000
+
+class TaprootKeySpendTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+        self.extra_args = [['-whitelist=noban@127.0.0.1',
+                            '-allownonstdtxnconsensus=1',
+                            '-acceptnonstdtxn=1',
+                            f'-replayprotectionactivationtime={ACTIVATION_TIME}']]
+
+    def run_test(self):
+        node = self.nodes[0]
+        node.add_p2p_connection(P2PDataStore())
+        # Allocate as many UTXOs as are needed
+        num_utxos = sum(test_case['inputs']
+                        for test_case in TESTCASES
+                        if isinstance(test_case, dict))
+
+        value = 50 * 100_000_000
+        fee = 10_000
+
+        max_utxo_value = (value - fee) // num_utxos
+        private_keys = []
+        public_keys = []
+        spendable_outputs = []
+        executed_scripts = []
+        utxo_idx = 0
+        # Prepare UTXOs for the tests below
+        for test_case in TESTCASES:
+            if test_case == 'ENABLE_REPLAY_PROTECTION':
+                continue
+            for _ in range(test_case['inputs']):
+                private_key = ECKey()
+                private_key.generate()
+                private_keys.append(private_key)
+                public_key = private_key.get_pubkey()
+                public_keys.append(public_key)
+                utxo_value = max_utxo_value - utxo_idx * 100  # deduct 100*i coins for unique amounts
+                utxo_script = CScript([OP_SCRIPTTYPE, OP_1, public_key.get_bytes()])
+                executed_scripts.append(utxo_script)
+                spendable_outputs.append(CTxOut(utxo_value, utxo_script))
+                utxo_idx += 1
+
+        anyonecanspend_address = node.decodescript('51')['p2sh']
+        burn_address = node.decodescript('00')['p2sh']
+        p2sh_script = CScript([OP_HASH160, bytes(20), OP_EQUAL])
+        node.generatetoaddress(1, anyonecanspend_address)
+        node.generatetoaddress(100, burn_address)
+
+        # Build and send fan-out transaction creating all the UTXOs
+        block_hash = node.getblockhash(1)
+        coin = int(node.getblock(block_hash)['tx'][0], 16)
+        tx_fan_out = CTransaction()
+        tx_fan_out.vin.append(CTxIn(COutPoint(coin, 0), CScript([b'\x51'])))
+        tx_fan_out.vout = spendable_outputs
+        tx_fan_out.rehash()
+
+        # Broadcast fan-out tx
+        node.p2p.send_txs_and_test([tx_fan_out], node)
+
+        utxo_idx = 0
+        key_idx = 0
+        for test_case in TESTCASES:
+            if test_case == 'ENABLE_REPLAY_PROTECTION':
+                node.setmocktime(ACTIVATION_TIME)
+                node.generatetoaddress(11, burn_address)
+                continue
+            # Build tx for this test, will broadcast later
+            tx = CTransaction()
+            num_inputs = test_case['inputs']
+            spent_outputs = spendable_outputs[:num_inputs]
+            del spendable_outputs[:num_inputs]
+            assert len(spent_outputs) == num_inputs
+            total_input_amount = sum(output.nValue for output in spent_outputs)
+            max_output_amount = (total_input_amount - fee) // test_case['outputs']
+            for i in range(test_case['outputs']):
+                output_amount = max_output_amount - i * 77
+                output_script = CScript([OP_HASH160, i.to_bytes(20, 'big'), OP_EQUAL])
+                tx.vout.append(CTxOut(output_amount, output_script))
+            for _ in range(test_case['inputs']):
+                tx.vin.append(CTxIn(COutPoint(int(tx_fan_out.hash, 16), utxo_idx), CScript()))
+                utxo_idx += 1
+            if test_case.get('sigs', None) is not None:
+                for i, sig in enumerate(test_case['sigs']):
+                    tx.vin[i].scriptSig = CScript([sig])
+                    key_idx += 1
+            else:
+                for i, sig_hash_type in enumerate(test_case['sig_hash_types']):
+                    # Compute sighash for this input; we sign it manually using sign_ecdsa/sign_schnorr
+                    # and then broadcast the complete transaction
+                    sighash = SignatureHashBIP341(
+                        tx_to=tx,
+                        spent_utxos=spent_outputs,
+                        sig_hash_type=sig_hash_type,
+                        input_index=i,
+                    )
+                    if test_case.get('sig', None) is not None:
+                        signature = test_case['sig']
+                    if test_case.get('ecdsa', False):
+                        signature = private_keys[key_idx].sign_ecdsa(sighash)
+                    else:
+                        signature = private_keys[key_idx].sign_schnorr(sighash)
+                    signature += bytes([test_case.get('suffix', sig_hash_type & 0xff)])
+                    # Build correct scriptSig
+                    tx.vin[i].scriptSig = CScript([signature])
+                    key_idx += 1
+            # Broadcast transaction and check success/failure
+            tx.rehash()
+            if 'error' not in test_case:
+                node.p2p.send_txs_and_test([tx], node)
+            else:
+                node.p2p.send_txs_and_test([tx], node, success=False, reject_reason=test_case['error'])
+
+
+if __name__ == '__main__':
+    TaprootKeySpendTest().main()
-- 
2.39.5 (Apple Git-154)

